Tasks/ContainerBuildV0/src/buildctl.ts

```diff
@@ -0,0 +1,103 @@
+    var buildctlPath = await utils.downloadBuildctl(stableBuildKitVersion);
+
+    // prepend the tools path. instructs the agent to prepend for future tasks
+    if (!process.env['PATH'].startsWith(path.dirname(buildctlPath))) {
+        toolLib.prependPath(path.dirname(buildctlPath));
+    }
+}
+
+async function verifyBuildctl() {
+    var buildctlToolPath = tl.which("buildctl", true);
+    if(buildctlToolPath == "")
+        await configureBuildctl();
+    tl.debug(tl.loc("VerifyBuildctlInstallation"));
+    
+    var buildctlTool = tl.tool(buildctlToolPath);
+    var executionOption : tr.IExecOptions = <any> {
+        silent: true
+    };
+
+    buildctlTool.arg("--help");
+    buildctlTool.exec(executionOption);
+}
+
+export async function buildctlBuildAndPush() {
+
+    await verifyBuildctl();
+
+    await utils.getBuildKitPod();
+
+    let tags = tl.getDelimitedInput("tags", "\n");
+    let endpointId = tl.getInput("dockerRegistryServiceConnection");
+    let registryAuthenticationToken: RegistryAuthenticationToken = getDockerRegistryEndpointAuthenticationToken(endpointId);
+
+    // Connect to any specified container registry
+    let connection = new ContainerConnection();
+    connection.open(null, registryAuthenticationToken, true, false);
+    let repositoryName = tl.getInput("repository");
+    if (!repositoryName) {
+        tl.warning("No repository is specified. Nothing will be pushed.");
+    }
+
+    let imageNames: string[] = [];
+    if (tl.getInput("dockerRegistryServiceConnection")) {
+        let imageName = connection.getQualifiedImageName(repositoryName, true);
+        if (imageName) {
+            imageNames.push(imageName);
+        }
+    }
+
+    var dockerfilefolder = tl.getInput("Dockerfile", true);
+    if(dockerfilefolder == "Dockerfile")
+    {
+        dockerfilefolder = ".";
+    }
+    else {
+        var index = dockerfilefolder.lastIndexOf("Dockerfile");
+        dockerfilefolder = dockerfilefolder.substring(0,index);
+        tl.debug("Dockerfilefolder path: "+dockerfilefolder);
+
+    }
+    var contextarg = "--local=context="+tl.getInput("buildContext", true);
+    var dockerfilearg = "--local=dockerfile="+dockerfilefolder;
+    var buildctlToolPath = tl.which("buildctl", true);
+    var buildctlTool = tl.tool(buildctlToolPath);
+
+    buildctlTool.arg("build");
+    buildctlTool.arg('--frontend=dockerfile.v0');
+    buildctlTool.arg(contextarg);
+    buildctlTool.arg(dockerfilearg);
+    if (imageNames && imageNames.length > 0) {
+        imageNames.forEach(imageName => {
+            if (tags && tags.length > 0) {
+                tags.forEach(async tag => {
+                    buildctlTool.arg(`--output=type=image,name=${imageName}:${tag},push=true`);
+                    buildctlTool.exec();
+                })
+            }
+            else {
+                buildctlTool.arg(`--output=type=image,name=${imageName}:latest,push=true`);
```

Comment:
Should not be using latest if tags input is left unspecified. The default value of $(Build.BuildId) should be used. In that case, will this else clause ever be executed?

Reply:
else would be executed if user makes the tag input as completely blank. Had added the latest tag as same is done in docker task

Reply:
Let me know if still we want the tag to be $(Build.BuildId) if the field is blank

Reply:
If tags field empty don't  add tag.

Reply:
removed tag if field is empty